### DISCUSSION TITLE
Resource Limit QOS

### DISCUSSION THREAD

**Discussion by ussefdesouky** [eProsima/Fast-DDS#4200] [Support]
Hello Fast-DDS Communtiy,

I created an idl message that includes the following:
> struct chatter{
    uint64 timestamp; // 8bytes
    uint64 id; // 8 bytes
    octet[10] message; // 10bytes
};

and I need to a small brief on what are samples and what do they present, secondly how to calculate the resource limit qos wrt to this message and the effect on putting very small values and infinite values on the hw resources and on the data writers and data readers.

1. max_samples
2. max_instances
3. max_samples/instance
4. allocated_samples
5. extra_samples

Thanks

**Comment by JesusPoderoso** [comment]
Hi @ussefdesouky, thanks for using Fast DDS.
The [Resource Limits QoS Policy section](https://fast-dds.docs.eprosima.com/en/latest/fastdds/dds_layer/core/policy/standardQosPolicies.html#resourcelimitsqospolicy) of the documentation describes each of the policies you have listed.

To understand it properly, let's define briefly what are samples and topic instances.

A sample is a data-type message. Having the below IDL which represents a _Temperature_ data-type, a sample is a sent/received piece of information structured in that data-type.

An [instance](https://fast-dds.docs.eprosima.com/en/latest/fastdds/dds_layer/topic/instances.html) is a logical separation in a Topic. Topic instances share the data-type, but include a [key](https://fast-dds.docs.eprosima.com/en/latest/fastdds/dds_layer/topic/instances.html) to distinguish which is the related instance.
My friend and collaborator @EduPonz used to describe it with the following example:

_"Let's say there is a Temperature topic where the data-types are temperature values represented in different units:"_
```
struct Temperature
{
	unsigned short celsius;
	unsigned short fahrenheit;
	unsigned short kelvin;
	@Key char[10] location;
};
```
_"So a Datawriter can publish Temperature samples in different locations. Instances allow to use the same DataWriter to publish each of them. The same DataWriter publishes kitchen temperature, but also living room temperature. Kitchen and living room share data type, but they have that logical distinction."_

Then, the resource limits define the application behavior regarding the samples or instances management.

> 1. max_samples

Controls the maximum number of samples that the DataWriter or DataReader can manage across all the instances associated with it. In other words, it represents the maximum samples that the middleware can store for a DataReader or DataWriter. **Value 0 means infinite resources.**

> 2. max_instances

Controls the maximum number of instances that a DataWriter or DataReader can manage. **Value 0 means infinite resources.**

> 3. max_samples/instance

Controls the maximum number of samples within an instance that the DataWriter or DataReader can manage. **Value 0 means infinite resources.**

> 4. allocated_samples

States the number of samples that will be allocated on initialization.

> 5. extra_samples

States the number of extra samples that will be allocated on the pool, so the maximum number of samples on the pool will be `max_samples` plus `extra_samples`. These extra samples act as a reservoir of samples even when the history is full.
